torchci/lib/metricUtils.ts

```diff
@@ -0,0 +1,193 @@
+) {
+  if (!(jobName in failures)) {
+    failures[jobName] = {
+      [JobAnnotation.BROKEN_TRUNK]: 0,
+      [JobAnnotation.INFRA_BROKEN]: 0,
+      [JobAnnotation.TEST_FLAKE]: 0,
+    };
+  }
+
+  return failures[jobName];
+}
+
+function updateBrokenInfraCount(
+  jobName: string,
+  count: number,
+  failures: { [jobName: string]: { [t: string]: number } }
+) {
+  if (count === 0) {
+    return;
+  }
+
+  const failure = getFailureByJobName(jobName, failures);
+  failure[JobAnnotation.INFRA_BROKEN] += count;
+}
+
+function updateFailureCount(
+  jobName: string,
+  count: number,
+  failures: { [jobName: string]: { [t: string]: number } },
+  is_broken_trunk: boolean
+) {
+  if (count === 0) {
+    return;
+  }
+
+  const failure = getFailureByJobName(jobName, failures);
+  failure[
+    is_broken_trunk ? JobAnnotation.BROKEN_TRUNK : JobAnnotation.TEST_FLAKE
+  ] += count;
+}
+
+export function approximateSuccessByJobName(
+  // The data from Rockset is sorted by time DESC, so newer commits come first
+  data?: JobsPerCommitData[]
+) {
+  const successesByJobName: { [success: string]: number } = {};
+
+  if (data === undefined || data === null) {
+    return successesByJobName;
+  }
+
+  data.forEach((commit: JobsPerCommitData) => {
+    const successes = new Set(
+      commit.successes.filter(
+        (n) => n !== null && n !== undefined && n.length > 0
+      )
+    );
+
+    // Iterate though all the successes in the commit and aggregate them by name
+    successes.forEach((success: string) => {
+      if (!(success in successesByJobName)) {
+        // Start the count
+        successesByJobName[success] = 0;
+      }
+
+      successesByJobName[success] += 1;
+    });
+  });
+
+  return successesByJobName;
+}
+
+export function approximateFailureByType(
+  // The data from Rockset is sorted by time DESC, so newer commits come first
+  data?: JobsPerCommitData[],
+  broken_trunk_threshold: number = BROKEN_TRUNK_THRESHOLD,
+  outage_threshold: number = OUTAGE_THRESHOLD
+) {
+  const failuresByTypes: { [failure: string]: { [t: string]: number } } = {};
+
+  if (data === undefined || data === null) {
+    return failuresByTypes;
+  }
+
+  const failuresCount: { [failure: string]: number } = {};
+  data.forEach((commit: JobsPerCommitData) => {
+    const failures = new Set(
+      commit.failures.filter(
+        (n) => n !== null && n !== undefined && n.length > 0
+      )
+    );
+
+    // Iterate though all the failures in the commit and aggregate them by name
+    failures.forEach((failure: string) => {
+      if (!(failure in failuresCount)) {
+        // Start the count
+        failuresCount[failure] = 0;
+      }
+
+      failuresCount[failure] += 1;
+    });
+
+    // Check if the job still fail in this commit
+    Object.keys(failuresCount).forEach((failure: string) => {
+      if (failures.has(failure)) {
+        // Count the commit as part of an outage
+        if (failures.size >= outage_threshold) {
```

Comment:
Any idea how accurate this type of an "outage detector" would be?  My gut feeling is "not very", but perhaps looking at past data may show otherwise.

Another option could be to include data from the manual job label mapping to detect which jobs failed due to an outage